Principal curvatures

ejpaul opened the pull request:
This PR implements second derivatives of the position vector on a Surface wrt the angles. This is used to evaluate the first and second fundamental form and the principal curvatures, which can serve as regularization terms in fixed-boundary optimization. Derivatives of these metrics are computed wrt the Surface coefficients.

landreman commented:
Thanks for this contribution @ejpaul. A few notes:

If I'm understanding correctly, it looks like the new `surface_curvature` function returns the mean curvature, Gaussian curvature, and the two principal curvatures. This should be documented somewhere. The plural name `surface_curvatures` could help indicate that it is not just the mean curvature or just the Gaussian curvature that is returned.

Generally, let's add all the new surface quantities (`surface_curvature`, `first_fund_form`, `dgammadash1dash1_by_dcoeff`, etc) to the list in `docs/source/geo.rst`.

This is a question also for @florianwechsung : Shouldn't each `.def` line in `python_surfaces.cpp` have a docstring, like in `python_magneticfield.cpp`?

For the docstring of the new `PrincipalCurvature` class, a little more could be said about the big picture and why one would use this objective. Is the point of this objective to avoid concave regions? Or something else?

A unit test that checks the Gauss-Bonnet theorem would be cool. Ages ago I used this to test some routines that computed the curvatures.

Is it necessary to delete the `input.*` files? Even if tests presently use the associated `wout` files but not the `input` files, it could be useful to keep the `input` files in the repository so the provenance of the `wout` files is documented.

ejpaul commented:
> Thanks for this contribution @ejpaul. A few notes:
> 
> If I'm understanding correctly, it looks like the new `surface_curvature` function returns the mean curvature, Gaussian curvature, and the two principal curvatures. This should be documented somewhere. The plural name `surface_curvatures` could help indicate that it is not just the mean curvature or just the Gaussian curvature that is returned.
> 
> Generally, let's add all the new surface quantities (`surface_curvature`, `first_fund_form`, `dgammadash1dash1_by_dcoeff`, etc) to the list in `docs/source/geo.rst`.
> 
> This is a question also for @florianwechsung : Shouldn't each `.def` line in `python_surfaces.cpp` have a docstring, like in `python_magneticfield.cpp`?
> 
> For the docstring of the new `PrincipalCurvature` class, a little more could be said about the big picture and why one would use this objective. Is the point of this objective to avoid concave regions? Or something else?
> 
> A unit test that checks the Gauss-Bonnet theorem would be cool. Ages ago I used this to test some routines that computed the curvatures.
> 
> Is it necessary to delete the `input.*` files? Even if tests presently use the associated `wout` files but not the `input` files, it could be useful to keep the `input` files in the repository so the provenance of the `wout` files is documented.

I believe I have addressed all of your comments. I have added docstrings to `python_surfaces.cpp` for the definitions related to this pull request.

florianwechsung commented on src/simsopt/field/boozermagneticfield.py:
```diff
@@ -361,61 +361,90 @@ def __init__(self, equil, order, mpol=32, ntor=32, N=None, enforce_vacuum=False,
             if self.booz.need_to_run_code:
                 self.booz.run()
 
+        self.asym = self.booz.bx.asym  # Bool for stellarator asymmetry
```
AFAIK, everywhere else in the code we store a bool called `stellsym` that is true when symmetry is satisfies. For consistency it would be nice to change this to
```
self.stellsym = not self.booz.bx.asym
```

ejpaul replied:
done

florianwechsung commented on src/simsopt/geo/surfaceobjectives.py:
```diff
@@ -141,6 +141,70 @@ def d2J_by_dsurfacecoefficientsdsurfacecoefficients(self):
+        self._J = None
+        self._dJ = None
+
+    def compute(self):
```
What is the reason for this structure? Isn't this very expensive if someone just wants to compute the objective?

ejpaul replied:
You are correct. I updated the structure.

florianwechsung commented on src/simsoptpp/surfacerzfourier.cpp:
```diff
@@ -606,6 +824,61 @@ void SurfaceRZFourier<Array>::dgammadash2_by_dcoeff_impl(Array& data) {
+                    if(m==0 && n<=0) continue;
+                    data(k1, k2, 0, counter) = 0;
+                    data(k1, k2, 1, counter) = 0;
+                    // data(k1, k2, 2, counter) = 2*M_PI*m * cos(m*theta-n*nfp*phi);
```
There are a few commented lines here. Can these be removed?

florianwechsung commented on src/simsoptpp/surfacexyzfourier.cpp:
```diff
@@ -115,6 +115,111 @@ void SurfaceXYZFourier<Array>::gammadash2_impl(Array& data) {
+                    // ydash += xhatdash * sin(phi) + yhatdash * cos(phi);
+                    ydd += xhatdd * sin(phi) + yhatdd * cos(phi);
+                    // zdash += -get_coeff(2, true , m, i) * (m) * sin(m*theta-n*nfp*phi) + get_coeff(2, false, m, i) * m * cos(m*theta-n*nfp*phi);
+                    zdd += -get_coeff(2, true , m, i) * (m) * (m) * cos(m*theta-n*nfp*phi) - get_coeff(2, false, m, i) * m * m * sin(m*theta-n*nfp*phi);
```
commented lines - can they go?

florianwechsung commented on src/simsoptpp/surfacexyzfourier.cpp:
```diff
@@ -273,6 +378,216 @@ void SurfaceXYZFourier<Array>::dgammadash2_by_dcoeff_impl(Array& data) {
+                                                       - sin(m*theta-n*nfp*phi) * cos(phi);
+                        }
+                        else if(d == 2) {
+                            // data(k1, k2, 2, counter) = (-n*nfp)*cos(m*theta-n*nfp*phi);
```
as above

florianwechsung commented on src/simsoptpp/surfacexyzfourier.cpp:
```diff
@@ -273,6 +378,216 @@ void SurfaceXYZFourier<Array>::dgammadash2_by_dcoeff_impl(Array& data) {
+                                                       - m * cos(m*theta-n*nfp*phi) * sin(phi);
+                        }
+                        else if(d == 2) {
+                            // data(k1, k2, 2, counter) = (-n*nfp)*cos(m*theta-n*nfp*phi);
```
as above

ejpaul replied:
I removed the unneeded comments

florianwechsung reviewed (commented):
Thank you @ejpaul for this contribution! I did a first pass over the code. Unfortunately we don't have automated code coverage checks for the C++, so I'll still have to manually check whether all the new C++ code is covered by tests before approving.

landreman reviewed (commented):
Thanks for the updates.

ejpaul commented on docs/source/simsopt.rst:
```diff
@@ -7,14 +7,12 @@ Subpackages
    simsopt.objectives
    simsopt.solve
    simsopt.util
-   simsoptpp
```
All of these changes to the .rst files occurred when I ran "sphinx-apidoc -f -o source ../src/simsopt" to update the docs. I have reverted this change.

ejpaul commented on docs/source/simsopt.rst:
```diff
@@ -7,14 +7,12 @@ Subpackages
 .. toctree::
    :maxdepth: 4
 
-   simsopt._core
```
All of these changes to the .rst files occurred when I ran "sphinx-apidoc -f -o source ../src/simsopt" to update the docs. I have reverted these changes.

florianwechsung commented:
This looks good to me now.

florianwechsung reviewed (approved):
Thank you!
